### ISSUE TITLE
Calling `.each(0, 'x')` in `on_init()` does not work?

### ISSUE THREAD

**Issue by Peque** [opensistemas-hub/osbrain#206]
This method is executed in `Agent.__init__()`.

Currently, if we execute `.each(0.5, 'x')`, then the agent will crash as the `Agent._loopback()` method dos not support operation when the agent is not running (i.e.: will raise a `NotImplementedError` exception).

We have two options:

- Either we support that operation (i.e.: we return `None` instead of raising an exception). This way messages will be discarded while the agent is not running.
- We create a new method `.before_running()` or `.configure()` which would allow the user to execute code just before entering the `self.loop()` (including `.each(...)` calls). This method should be documented and, maybe, preferred over `.on_init()`, which could be used by advanced users for some special cases.

I think I would rather go with the second option, naming the method `.configure()` or something else but simple/intuitive for the user. @Flood1993 What do you think? Maybe there are other options than those two?

------------

Code to reproduce the error:

```python
import time

from osbrain import Agent
from osbrain import run_agent
from osbrain import run_nameserver


class MyAgent(Agent):
    def on_init(self):
        self.each(0.5, 'x')

    def x(self):
        print('asdf')


ns = run_nameserver()
run_agent('asdf', base=MyAgent)
time.sleep(1.5)
ns.shutdown()
```

**Comment by ocaballeror** [comment]
Trying to  think about the implementation of the second option.

If we are going to be able to execute code before the agent is running, we will have to modify the `_loopback` method anyway, so maybe we should add a parameter to that one. I'm thinking about something like this:

```python
def _loopback(self, header, data=None, ignore_state=False):
  if not self.running:
      if ignore_state:
          return None
      raise NotImplementedError()
    ...
```

Is this in line with what you were thinking?

**Comment by Peque** [comment]
@ocaballeror Not really I think. The idea is to create a new method `.configure()` that is executed before entering `self.loop()`.

This approach should not require modifying `_loopback()`. Just add the method call before `self.loop()` and implement a dummy `.configure()` method that does nothing. Is up to the user to implement this method in order to execute code for configuring the agent.

Probably we should change the documentation to use `.configure()` everywhere instead of `.on_init()` and mention that there is also `.on_init()` method if the user wants to execute code even before the agent is running, just after initialization.

**Comment by ocaballeror** [comment]
But that would not fix the problem, would it?

I understand how the `.configure()` method should work, but if you want to call `.each()`  inside it, you would get the same error as you do now with the code you posted at the start, since the agent won't be running at that point.

If we are to be able to execute code before the agent is running, we will have to modify `_loopback()` in some way.

**Comment by Peque** [comment]
@ocaballeror I think it would fix it, yeah. :wink:

The idea would be to execute `.configure()` just before `.loop()`. We can consider the `.configure()` agent state as running as events will be processed right after (just like with any other `.loop()` iteration, only we are out of the loop and only execute that once).

**Comment by ocaballeror** [comment]
Oh, OK, I see what you mean now. `.configure()` should be called after setting the `running` attribute to True, but before calling `loop()`, thus it will be effectively running when `configure` is called.

Took me a while to get it, but OK. I'll start working on it.

**Comment by Peque** [comment]
Closed by #229.
